### PR TITLE
feat: initial commit

### PR DESCRIPTION
**Pull request by poolpitako**

### CODE REVIEW

File: contracts/RouterStrategy.sol
Comment (fp-crypto) [inline contracts/RouterStrategy.sol:6]: I think this is importing version `0.4.2` right now.

File: contracts/RouterStrategy.sol
Comment (fp-crypto) [inline contracts/RouterStrategy.sol:148]: Could the amount to liquidate be greater than the totalDebt? Maybe ask for `type(uint256).max`?

Comment (poolpitako) [reply contracts/RouterStrategy.sol:148]: I just checked this.

Vault does the check in https://github.com/yearn/yearn-vaults/blob/master/contracts/Vault.vy#L1081
I also can't do max int because `_investmentTokenToYShares()` does a `mul()`.

Comment (fp-crypto) [reply contracts/RouterStrategy.sol:148]: Could `vault.strategies(address(this)).totalDebt` < `_investmentTokenToYShares(yVault.balanceOf(address(this)))`?

Comment (fp-crypto) [reply contracts/RouterStrategy.sol:148]: Could `vault.strategies(address(this)).totalDebt` < `valueOfInvestment()`?

Comment (poolpitako) [reply contracts/RouterStrategy.sol:148]: yep, but in that case it will be harvested when `harvest()` is called

Comment (fp-crypto) [reply contracts/RouterStrategy.sol:152]: 👍

Comment (plutofeather) [reply contracts/RouterStrategy.sol:152]: I was thinking the same thing. So there's a chance that all positions would be liquidated and there will still be some profit left in the underlying vault. Wouldn't that make PPS skyrocket on the next harvest?

File: contracts/RouterStrategy.sol
Comment (plutofeather) [inline contracts/RouterStrategy.sol:194]: Is there some mechanism to make sure there are no losses when withdrawn during the sandwich protection window? More just curious how it works here.

File: contracts/RouterStrategy.sol
Comment (poolpitako) [reply contracts/RouterStrategy.sol:152]: Pretty sure it will not but let me write a test.
I will:
 - harvest
 - send profit to the routed vault
 - set emergency exit
 - harvest
 
 and see what happens 😨

File: contracts/RouterStrategy.sol
Comment (poolpitako) [reply contracts/RouterStrategy.sol:194]: Good point. 
No protection, only us calling it on the right time.

If we set deposits to 0, we can do it every week and it's pretty safe.

File: contracts/RouterStrategy.sol
Comment (hsukev) [inline contracts/RouterStrategy.sol:88]: I think `_profit` might be double counted here since `liquidatePosition` takes into account the existing balance.
i.e  if `debtOutstanding = 10000` and `_profit = 100` 
-> `liquidatePosition(10000)`
-> `toWithdraw = 10000 - 100 = 9900`
-> `totalAssets = 10000`, `liquidatedAmount = 10000`
-> `(_amountFreed, _loss) = (10000,0)`
-> `_debtPayment = 10000`
now you have an extra 100 reported as profit but only 10000 of want free

Comment (hsukev) [reply contracts/RouterStrategy.sol:88]: one solution would be to just withdraw the amount requested, without taking into account existing balance (`10000 - 100` part)

File: contracts/RouterStrategy.sol
Comment (hsukev) [inline contracts/RouterStrategy.sol:120]: nit: ` = balanceOfWant()`

File: contracts/RouterStrategy.sol
Comment (hsukev) [reply contracts/RouterStrategy.sol:152]: isn't it better here to just `yVault.withdraw(yVault.balanceOf(address(this)), address(this), maxLoss)`? Otherwise you're relying on good record keeping by going through the `liquidatePosition` route. Someone could theoretically airdrop some want into this strat and cause some of the yVault shares to remain locked up.

File: contracts/RouterStrategy.sol
Comment (orbxball) [reply contracts/RouterStrategy.sol:88]: another alternative w/o changing too much is to do `liquidatePosition(_debtOutstanding.add(balanceOfWant))`

Comment (hsukev) [reply contracts/RouterStrategy.sol:88]: yup that works too

File: contracts/RouterStrategy.sol
Comment (poolpitako) [reply contracts/RouterStrategy.sol:120]: fixed. thanks

File: contracts/RouterStrategy.sol
Comment (poolpitako) [reply contracts/RouterStrategy.sol:152]: found a bug doing this test. well done @plutofeather!

File: contracts/RouterStrategy.sol
Comment (poolpitako) [reply contracts/RouterStrategy.sol:88]: @hsukev good catch, I am going to go with @orbxball recommendation.

File: contracts/RouterStrategy.sol
Comment (poolpitako) [reply contracts/RouterStrategy.sol:152]: Will push new tests soon.

File: contracts/RouterStrategy.sol
Comment (jmonteer) [reply contracts/RouterStrategy.sol:90]: wouldn't it be better to use @orbxball 's but instead of balanceOfWant use _profit?

as in `liquidatePosition(_debtOutstanding.add(_profit))`

**Review by jmonteer — APPROVED** [review]

**Review by orbxball — APPROVED** [review]

**Review by plutofeather — APPROVED** [review]

File: contracts/RouterStrategy.sol
Comment (poolpitako) [reply contracts/RouterStrategy.sol:92]: will push an update to `prepareReturn`. thanks!

File: contracts/RouterStrategy.sol
Comment (poolpitako) [reply contracts/RouterStrategy.sol]: added in the latest version.

**Review by hsukev — APPROVED** [review]
LGTM

**Review by retro-wavey — APPROVED** [review]

File: contracts/RouterStrategy.sol
Comment (poolpitako) [reply contracts/RouterStrategy.sol:82]: can you give an example? it would be considered in `liquidatePosition()`

File: contracts/RouterStrategy.sol
Comment (poolpitako) [reply contracts/RouterStrategy.sol]: it looks like you are viewing an old diff. The change is there:
https://github.com/poolpitako/RouterStrategy/pull/1/files#diff-3991d5d1f579e1182ca2ef263e785f30f78e373b22cdac9b4dfb6f83ae9cf3bdR91-R105

will resolve this one.

File: contracts/RouterStrategy.sol
Comment (storming0x) [inline contracts/RouterStrategy.sol:228]: I guess its just a comment, Depending on the version of the vault pricePerShare it may be beneficial to wait for harvest to go past the locked profit release window after a harvest

File: contracts/RouterStrategy.sol
Comment (storming0x) [inline contracts/RouterStrategy.sol:199]: Change it to a local modifier, onlyManagement() (we should add it to base Strategy)

```
   modifier onlyManagement() {
        require(
            msg.sender == governance() || msg.sender == vault.management(),
            "!authorized"
        );
        _;
    }
    
    ```

Comment (storming0x) [reply contracts/RouterStrategy.sol:199]: i added an issue for it but in the meantime we can do customn here

https://github.com/yearn/yearn-vaults/issues/413

Comment (poolpitako) [reply contracts/RouterStrategy.sol:199]: done!

File: contracts/RouterStrategy.sol
Comment (retro-wavey) [reply contracts/RouterStrategy.sol:82]: Consider when 042 vault suffers a loss/hack. 

Pps would still show as healthy on old vault A, but a withdraw would suffer a loss when pulling from new vault.

Could block withdraws from web page if maxLoss is exceeded.

File: contracts/RouterStrategy.sol
Comment (poolpitako) [reply contracts/RouterStrategy.sol:228]: that's an interesting protection we should do. make sure sleeping 8hs before the tx doesn't affect the pps.

File: contracts/RouterStrategy.sol
Comment (jmonteer) [reply contracts/RouterStrategy.sol:82]: very good point @retro-wavey 

thinking about this, some thoughts: 
1. if we consider vaults as up-only, having a loss should be considered as a special case
2. when a hack happens (rare thing), either we prefer to hold people from withdrawing (although is still posible if they take the losses, setting maxLoss higher) or we will reimburse, making pps to go up and erasing the issue after applying a solution. In any case, this "bug" would buy us time
3. We can set maxLoss to 10_000 from this vault to avoid being stopped from withdrawing from newer vault
4. Adding the else clause would mean that if we harvest we take the loss no matter what happens afterward

im not sure what would I do here but for sure I have pros and cons for both approaches

IMO it is cleaner if we add the else clause but as it is it would work for sure. 

taking into account that in any case we need to harvest (either realising the loss through a lower debt ratio) or harvesting to report the loss, it may be simpler/safer to not add the else clause (as we would report losses 100% on purpose)

Comment (retro-wavey) [reply contracts/RouterStrategy.sol:82]: Good comments.

Personally I am fine either way. 

Poolpi, you have my LGTM.

Comment (poolpitako) [reply contracts/RouterStrategy.sol:82]: cool, this is great. will mark as resolved and merge.

**Pull request merged**
